File: src/env.ts
export function envInt(name: string, fallback: number, min = 1): number {
  const raw = process.env[name];
  if (!raw) {
    return fallback;
  }
  const n = Number(raw);
  return Number.isFinite(n) && n >= min ? Math.floor(n) : fallback;
}

export function envFirstString(names: string[], fallback: string): string {
  for (const name of names) {
    const value = process.env[name];
    if (value) {
      return value;
    }
  }
  return fallback;
}

export function envOptionalInt(names: string[], min = 1): number | undefined {
  for (const name of names) {
    const raw = process.env[name];
    if (!raw) {
      continue;
    }
    const n = Number(raw);
    if (Number.isFinite(n) && n >= min) {
      return Math.floor(n);
    }
  }
  return undefined;
}

export const TAKARA_API_KEY_ENV = "TAKARA_API_KEY";

const MCP_KEY_PLACEHOLDER = "$" + "{TAKARA_API_KEY}";
const TAKARA_API_KEY_PLACEHOLDERS = new Set([MCP_KEY_PLACEHOLDER, "$TAKARA_API_KEY"]);

/** True when the value is a real key, not empty or an unexpanded MCP env placeholder. */
export function isUsableTakaraApiKey(value: string | undefined): boolean {
  const key = value?.trim() ?? "";
  if (!key) {
    return false;
  }
  return !TAKARA_API_KEY_PLACEHOLDERS.has(key);
}

/** Drop placeholder values so credentials.json can supply the key for MCP. */
export function normalizeTakaraApiKeyEnv(): void {
  if (!isUsableTakaraApiKey(process.env[TAKARA_API_KEY_ENV])) {
    delete process.env[TAKARA_API_KEY_ENV];
  }
}

export function hasTakaraApiKeyInEnv(): boolean {
  return isUsableTakaraApiKey(process.env[TAKARA_API_KEY_ENV]);
}

export function resolveEmbeddingApiKey(): string {
  const key = process.env[TAKARA_API_KEY_ENV]?.trim() ?? "";
  if (!isUsableTakaraApiKey(key)) {
    throw new Error(
      "Takara API key required. Run `miru setup`, or set TAKARA_API_KEY in your MCP server env or .env.local.",
    );
  }
  return key;
}

File: internal/env/env.go
package env

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

const TakaraAPIKeyEnv = "TAKARA_API_KEY"

func Int(name string, fallback int, minValue ...int) int {
	min := 1
	if len(minValue) > 0 {
		min = minValue[0]
	}
	raw := os.Getenv(name)
	if raw == "" {
		return fallback
	}
	n, err := strconv.ParseFloat(raw, 64)
	if err != nil || n < float64(min) {
		return fallback
	}
	return int(n)
}

func FirstString(names []string, fallback string) string {
	for _, name := range names {
		if value := os.Getenv(name); value != "" {
			return value
		}
	}
	return fallback
}

func OptionalInt(names []string, minValue ...int) (int, bool) {
	min := 1
	if len(minValue) > 0 {
		min = minValue[0]
	}
	for _, name := range names {
		raw := os.Getenv(name)
		if raw == "" {
			continue
		}
		n, err := strconv.ParseFloat(raw, 64)
		if err == nil && n >= float64(min) {
			return int(n), true
		}
	}
	return 0, false
}

func HasTakaraAPIKeyInEnv() bool {
	return strings.TrimSpace(os.Getenv(TakaraAPIKeyEnv)) != ""
}

func ResolveEmbeddingAPIKey() (string, error) {
	key := strings.TrimSpace(os.Getenv(TakaraAPIKeyEnv))
	if key == "" {
		return "", fmt.Errorf("Takara API key required. Run `miru setup`, or set TAKARA_API_KEY in your MCP server env or .env.local")
	}
	return key, nil
}
